src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp

```diff
@@ -0,0 +1,236 @@
+#include "Options/Options.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/ProtocolHelpers.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+class DataVector;
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+namespace GeneralizedHarmonic {
+/*!
+ * \brief Computes the generalized-harmonic upwind multipenalty boundary
+ * correction.
+ *
+ * This implements the upwind multipenalty boundary correction term
+ * \f$D_\alpha\f$. The general form is given by:
+ *
+ * \f{align*}{
+ *   \label{eq:pnpm upwind boundary term characteristics}
+ *   D_\beta =
+ *   T_{\beta\hat{\beta}}^{\mathrm{ext}}
+ *   \Lambda^{\mathrm{ext},-}_{\hat{\beta}\hat{\alpha}}
+ *   v^{\mathrm{ext}}_{\hat{\alpha}}
+ *   -T_{\beta\hat{\beta}}^{\mathrm{int}}
+ *   \Lambda^{\mathrm{int},-}_{\hat{\beta}\hat{\alpha}}
+ *   v^{\mathrm{int}}_{\hat{\alpha}},
+ * \f}
+ *
+ * Note that Eq. (6.3) of \cite Teukolsky2015ega is not exactly what's
+ * implemented since that boundary term does not consistently treat both sides
+ * of the interface on the same footing.
+ *
+ * For the first-order GH system the correction is:
+ *
+ * \f{align}{
+ *   \label{eq:pnpm upwind gh metric}
+ *   D_{g_{ab}} &= \tilde{\lambda}_{v^{g}}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},g}_{ab}
+ *                - \tilde{\lambda}_{v^{g}}^{\mathrm{int}}
+ *                v^{\mathrm{int},g}_{ab}, \\
+ *   \label{eq:pnpm upwind gh pi}
+ *   D_{\Pi_{ab}}
+ *              &= \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},+}_{ab} +
+ *                \tilde{\lambda}_{v^-}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},-}_{ab}\right)
+ *                + \tilde{\lambda}_{v^g}^\mathrm{ext}\gamma_2
+ *                v^{\mathrm{ext},g}_{ab}
+ *                \notag \\
+ *              &-\frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{int}}
+ *                v^{\mathrm{int},+}_{ab} +
+ *                \tilde{\lambda}_{v^-}^{\mathrm{int}}
+ *                v^{\mathrm{int},-}_{ab}\right)
+ *                - \tilde{\lambda}_{v^g}^\mathrm{int}\gamma_2
+ *                v^{\mathrm{int},g}_{ab} , \\
+ *   \label{eq:pnpm upwind gh phi}
+ *   D_{\Phi_{iab}}
+ *              &= \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},+}_{ab}
+ *                - \tilde{\lambda}_{v^-}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},-}_{ab}\right)n_i^{\mathrm{ext}}
+ *                + \tilde{\lambda}_{v^0}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},0}_{iab}
+ *                \notag \\
+ *              &-
+ *                \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{int}}
+ *                v^{\mathrm{int},+}_{ab}
+ *                - \tilde{\lambda}_{v^-}^{\mathrm{int}}
+ *                v^{\mathrm{int},-}_{ab}\right)n_i^{\mathrm{int}}
+ *                - \tilde{\lambda}_{v^0}^{\mathrm{int}}
+ *                v^{\mathrm{int},0}_{iab},
+ * \f}
+ *
```

Comment:
Just a note: I checked by hand that these equations follow from the original definition after plugging in the char fields and speeds.